File: LeetCode_ProblemSet/Problem2551.cpp
#include <vector>
#include <algorithm>

class Solution {
    public:
    long long putMarbles(std::vector<int>& weights, int k) {
        std::vector<int> costs(weights.size() - 1);
        for (int idx = 0; idx < weights.size() - 1; idx++) {
            costs[idx] = weights[idx] + weights[idx + 1];
        }
        std::sort(costs.begin(), costs.end());

        long long answer = 0;
        for (int i = 0; i < k - 1; i++) {
            answer += costs[costs.size() - 1 - i] - costs[i]; 
        }
        return answer;
    }
};

File: LeetCode_ProblemSet/Rust/src/Problem2551.rs
impl Solution {
    pub fn put_marbles(weights: Vec<i32>, k: i32) -> i64 {
        let mut costs = vec![0; weights.len() - 1 as usize];
        for idx in 0..weights.len() - 1 {
            costs[idx] = weights[idx] + weights[idx + 1];
        }
        costs.sort();

        let mut answer: i64 = 0;
        for i in 0..k as usize - 1 {
            answer += costs[costs.len() - 1 - i] as i64 - costs[i] as i64;
        }
        return answer;
    }
}